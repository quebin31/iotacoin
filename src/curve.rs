use std::ops::Add;

use crate::{Error, Result};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct EllipticCurve {
    a: isize,
    b: isize,
}

impl EllipticCurve {
    pub fn new(a: isize, b: isize) -> Self {
        Self { a, b }
    }

    pub fn contains(&self, x: isize, y: isize) -> bool {
        y * y == (x * x * x) + self.a * x + self.b
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Point {
    AtInfinity,
    Normal(isize, isize, EllipticCurve),
}

impl Point {
    pub fn new(x: isize, y: isize, curve: EllipticCurve) -> Result<Self> {
        if curve.contains(x, y) {
            Ok(Self::Normal(x, y, curve))
        } else {
            Err(Error::PointNotInTheCurve(x, y))
        }
    }

    pub fn at_infinity() -> Self {
        Self::AtInfinity
    }

    pub fn x(&self) -> Option<isize> {
        match *self {
            Point::AtInfinity { .. } => None,
            Point::Normal(x, _, _) => Some(x),
        }
    }

    pub fn y(&self) -> Option<isize> {
        match *self {
            Point::AtInfinity { .. } => None,
            Point::Normal(_, y, _) => Some(y),
        }
    }

    pub fn curve(&self) -> Option<EllipticCurve> {
        match *self {
            Point::AtInfinity => None,
            Point::Normal(_, _, curve) => Some(curve),
        }
    }

    pub fn same_curve(&self, other: &Self) -> bool {
        match (self.curve(), other.curve()) {
            (Some(curve1), Some(curve2)) => curve1 == curve2,
            _ => true, // One is a point at infinity
        }
    }

    pub fn is_point_at_inf(&self) -> bool {
        matches!(self, Self::AtInfinity)
    }
}

impl Add for Point {
    type Output = Result<Self>;

    fn add(self, rhs: Self) -> Self::Output {
        if !self.same_curve(&rhs) {
            return Err(Error::PointsNotInTheSameCurve);
        }

        match (self, rhs) {
            (Self::AtInfinity, _) => Ok(rhs),
            (_, Self::AtInfinity) => Ok(self),
            (Self::Normal(x1, y1, curve), Self::Normal(x2, y2, _)) => match (x1 == x2, y1 == y2) {
                // Same x axis, rhs is additive inverse of self and viceversa
                (true, false) => Ok(Self::at_infinity()),

                // Same x and y axis, self is equal to rhs
                (true, true) => {
                    if y1 == 0 {
                        return Ok(Self::at_infinity());
                    }

                    let slope = (3 * x1 * x1 + curve.a) / (2 * y1);
                    let x3 = slope * slope - 2 * x1;
                    let y3 = slope * (x1 - x3) - y1;

                    Self::new(x3, y3, curve)
                }

                // Different x axis, y axis doesn't matter in this case
                _ => {
                    let slope = (y2 - y1) / (x2 - x1);
                    let x3 = slope * slope - x1 - x2;
                    let y3 = slope * (x1 - x3) - y1;

                    Self::new(x3, y3, curve)
                }
            },
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::Result;

    #[test]
    fn not_in_curve() {
        let res = Point::new(-1, -2, EllipticCurve::new(5, 7));
        assert!(res.is_err());
    }

    #[test]
    fn equality() -> Result<()> {
        let curve = EllipticCurve::new(5, 7);
        let a = Point::new(-1, -1, curve)?;
        let b = Point::new(18, 77, curve)?;

        assert_eq!(a, a);
        assert_ne!(a, b);

        Ok(())
    }

    #[test]
    fn addition_with_inf() -> Result<()> {
        let curve = EllipticCurve::new(5, 7);
        let a = Point::new(-1, -1, curve)?;
        let inf = Point::at_infinity();

        assert_eq!(a.add(inf)?, a);
        assert_eq!(inf.add(a)?, a);

        Ok(())
    }

    #[test]
    fn addition_with_inverse() -> Result<()> {
        let curve = EllipticCurve::new(5, 7);
        let a = Point::new(-1, -1, curve)?;
        let b = Point::new(-1, 1, curve)?;

        assert_eq!(a.add(b)?, Point::at_infinity());
        assert_eq!(b.add(a)?, Point::at_infinity());

        Ok(())
    }

    #[test]
    fn addition_diff_points() -> Result<()> {
        let curve = EllipticCurve::new(5, 7);
        let a = Point::new(-1, -1, curve)?;
        let b = Point::new(2, 5, curve)?;
        let c = Point::new(3, -7, curve)?;

        assert_eq!(a.add(b)?, c);
        assert_eq!(b.add(a)?, c);

        Ok(())
    }

    #[test]
    fn addition_equal_points() -> Result<()> {
        let curve = EllipticCurve::new(5, 7);
        let a = Point::new(-1, -1, curve)?;
        let b = Point::new(18, 77, curve)?;

        assert_eq!(a.add(a)?, b);

        Ok(())
    }
}
