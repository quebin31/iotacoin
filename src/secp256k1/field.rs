use std::ops::{Add, Div, Mul, Sub};

use lazy_static::lazy_static;
use num_bigint::{BigInt, BigUint, Sign};
use num_integer::Integer;
use num_traits::{One, Pow, Zero};

lazy_static! {
    /// `secp256k1` prime = 2^256 - 2^32 - 977
    pub(crate) static ref PRIME: BigUint =
        biguint!("fffffffffffffffffffffffffffffffffffffffffffffffffffffffefffffc2f");
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct FieldElement(pub(crate) BigUint);

impl FieldElement {
    /// Build a new element in the S256 field
    pub fn new<U>(number: U) -> Self
    where
        U: Into<BigUint>,
    {
        Self(number.into() % &*PRIME)
    }

    /// Get the _additive inverse_ of this element.
    #[inline]
    pub fn add_inv(&self) -> Self {
        Self(&*PRIME - &self.0)
    }

    /// Get the _multiplicative inverse_ of this element.
    #[inline]
    pub fn mul_inv(&self) -> Self {
        // Fermat's little theorem
        self.pow(&*PRIME - 2usize)
    }
}

impl Zero for FieldElement {
    fn zero() -> Self {
        FieldElement(BigUint::zero())
    }

    fn is_zero(&self) -> bool {
        self.0.is_zero()
    }
}

impl One for FieldElement {
    fn one() -> Self {
        FieldElement(BigUint::one())
    }
}

impl<'a, E> Pow<E> for &'a FieldElement
where
    E: Into<BigInt>,
{
    type Output = FieldElement;

    fn pow(self, exp: E) -> Self::Output {
        let exp: BigInt = exp.into();
        let exponent = match exp.to_biguint() {
            Some(exp) => exp,
            None => {
                let prime = BigInt::from_biguint(Sign::Plus, &*PRIME - BigUint::one());
                exp.mod_floor(&prime).to_biguint().unwrap() // safe
            }
        };

        let number = self.0.modpow(&exponent, &*PRIME);
        FieldElement(number)
    }
}

impl<'a, 'b> Add<&'a FieldElement> for &'b FieldElement {
    type Output = FieldElement;

    fn add(self, rhs: &'a FieldElement) -> Self::Output {
        let number = (&self.0 + &rhs.0) % &*PRIME;
        FieldElement::new(number)
    }
}

impl<'a, 'b> Sub<&'a FieldElement> for &'b FieldElement {
    type Output = FieldElement;

    fn sub(self, rhs: &'a FieldElement) -> Self::Output {
        self.add(&rhs.add_inv())
    }
}

impl<'a, 'b> Mul<&'a FieldElement> for &'b FieldElement {
    type Output = FieldElement;

    fn mul(self, rhs: &'a FieldElement) -> Self::Output {
        let number = (&self.0 * &rhs.0) % &*PRIME;
        FieldElement::new(number)
    }
}

impl<'a, 'b> Div<&'a FieldElement> for &'b FieldElement {
    type Output = FieldElement;

    fn div(self, rhs: &'a FieldElement) -> Self::Output {
        self.mul(&rhs.mul_inv())
    }
}

impl<'a> Mul<usize> for &'a FieldElement {
    type Output = FieldElement;

    fn mul(self, rhs: usize) -> Self::Output {
        self.mul(FieldElement::new(rhs))
    }
}

impl Mul<usize> for FieldElement {
    type Output = FieldElement;

    fn mul(self, rhs: usize) -> Self::Output {
        Mul::mul(&self, rhs)
    }
}

forward_binop_impl!(for non-copyable FieldElement where Add does add);
forward_binop_impl!(for non-copyable FieldElement where Sub does sub);
forward_binop_impl!(for non-copyable FieldElement where Mul does mul);
forward_binop_impl!(for non-copyable FieldElement where Div does div);
