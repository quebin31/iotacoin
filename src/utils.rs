use std::cmp::Ordering;

use hmac::{Hmac, Mac};
use ripemd160::Ripemd160;
use sha2::{Digest, Sha256};

use crate::Error;

pub(crate) fn prepend_padding<A, T>(vec: A, size: usize, with: T) -> Result<Vec<T>, Error>
where
    T: Clone,
    A: Into<Vec<T>>,
{
    let mut vec: Vec<_> = vec.into();

    match vec.len().cmp(&size) {
        Ordering::Greater => Err(Error::OverflowPadding),
        Ordering::Equal => Ok(vec),
        Ordering::Less => {
            let mut result = vec![with; size - vec.len()];
            result.append(&mut vec);
            Ok(result)
        }
    }
}

pub(crate) fn strip_start<T>(arr: &[T], elem: T) -> &[T]
where
    T: Eq,
{
    let mut new_start = 0;

    while arr[new_start] == elem {
        new_start += 1;

        if new_start == arr.len() {
            break;
        }
    }

    &arr[new_start..]
}

pub fn hash160<B>(data: B) -> Vec<u8>
where
    B: AsRef<[u8]>,
{
    let hasher = Sha256::new();
    let digest = hasher.chain(data.as_ref()).finalize();

    let hasher = Ripemd160::new();
    let digest = hasher.chain(digest).finalize();

    digest.as_slice().to_vec()
}

pub fn hash256<B>(data: B) -> Vec<u8>
where
    B: AsRef<[u8]>,
{
    let mut hasher = Sha256::new();

    hasher.update(data.as_ref());
    let digest = hasher.finalize_reset();

    hasher.update(digest);
    let digest = hasher.finalize();

    digest.as_slice().to_vec()
}

pub(crate) trait Chain {
    fn chain(self, data: &[u8]) -> Self;
}

impl Chain for Hmac<Sha256> {
    fn chain(mut self, data: &[u8]) -> Self {
        self.update(data);
        self
    }
}
