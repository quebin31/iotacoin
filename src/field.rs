use num_traits::Pow;

use crate::traits::{IsZero, MayAdd, MayDiv, MayMul, MaySub};
use crate::utils::pow_mod;
use crate::Error;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct FieldElement {
    number: usize,
    prime: usize,
}

impl FieldElement {
    /// Build a new field element `number` on F_{prime}.
    pub fn new(number: usize, prime: usize) -> Result<Self, Error> {
        if number >= prime {
            Err(Error::InvalidFieldNumber(number, prime))
        } else {
            Ok(Self { number, prime })
        }
    }

    /// Get the _additive inverse_ of this element.
    #[inline]
    pub fn add_inv(self) -> Self {
        let number = self.prime - self.number;
        Self { number, ..self }
    }

    /// Get the _multiplicative inverse_ of this element.
    #[inline]
    pub fn mul_inv(self) -> Self {
        // Fermat's little theorem
        self.pow(self.prime - 2)
    }
}

impl IsZero for FieldElement {
    fn is_zero(&self) -> bool {
        self.number == 0
    }
}

impl Pow<usize> for FieldElement {
    type Output = Self;

    fn pow(self, exp: usize) -> Self::Output {
        let number = pow_mod(self.number, exp, self.prime);
        Self { number, ..self }
    }
}

impl Pow<isize> for FieldElement {
    type Output = Self;

    fn pow(self, exp: isize) -> Self::Output {
        // Rust's `%` operator is not the modulus operation.
        let exp = exp.rem_euclid(self.prime as isize - 1);
        self.pow(exp as usize)
    }
}

impl MayAdd for FieldElement {
    type Output = Self;
    type Error = Error;

    #[inline]
    fn may_add(self, rhs: Self) -> Result<Self::Output, Self::Error> {
        if self.prime != rhs.prime {
            Err(Error::InvalidFieldAddition)
        } else {
            let number = (self.number + rhs.number) % self.prime;
            Ok(Self { number, ..self })
        }
    }
}

impl MaySub for FieldElement {
    type Output = Self;
    type Error = Error;

    #[inline]
    fn may_sub(self, rhs: Self) -> Result<Self::Output, Self::Error> {
        if self.prime != rhs.prime {
            Err(Error::InvalidFieldSubstraction)
        } else {
            self.may_add(rhs.add_inv())
        }
    }
}

impl MayMul for FieldElement {
    type Output = Self;
    type Error = Error;

    #[inline]
    fn may_mul(self, rhs: Self) -> Result<Self::Output, Self::Error> {
        if self.prime != rhs.prime {
            Err(Error::InvalidFieldMultiplication)
        } else {
            // quebin31: what happens if we overflow?
            let number = (self.number * rhs.number) % self.prime;
            Ok(Self { number, ..self })
        }
    }
}

impl MayMul<usize> for FieldElement {
    type Output = Self;
    type Error = Error;

    #[inline]
    fn may_mul(self, rhs: usize) -> Result<Self::Output, Self::Error> {
        let this = Self::new(rhs, self.prime)?;
        this.may_mul(self)
    }
}

impl MayDiv for FieldElement {
    type Output = Self;
    type Error = Error;

    #[inline]
    fn may_div(self, rhs: Self) -> Result<Self::Output, Self::Error> {
        if self.prime != rhs.prime {
            Err(Error::InvalidFieldDivition)
        } else {
            self.may_mul(rhs.mul_inv())
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::Result;

    #[test]
    fn equality() -> Result<()> {
        let a = FieldElement::new(7, 13)?;
        let b = FieldElement::new(6, 13)?;

        assert_ne!(a, b);
        Ok(())
    }

    #[test]
    fn addition() -> Result<()> {
        let a = FieldElement::new(6, 13)?;
        let b = FieldElement::new(7, 13)?;
        let c = a.may_add(b)?;

        assert_eq!(c, FieldElement::new(0, 13)?);
        Ok(())
    }

    #[test]
    fn substraction() -> Result<()> {
        let a = FieldElement::new(9, 57)?;
        let b = FieldElement::new(29, 57)?;
        let c = a.may_sub(b)?;

        assert_eq!(c, FieldElement::new(37, 57)?);
        Ok(())
    }

    #[test]
    fn multiplication() -> Result<()> {
        let a = FieldElement::new(3, 13)?;
        let b = FieldElement::new(12, 13)?;
        let c = a.may_mul(b)?;

        assert_eq!(c, FieldElement::new(10, 13)?);
        Ok(())
    }

    #[test]
    fn exponentiation() -> Result<()> {
        use num_traits::Pow;

        let a = FieldElement::new(7, 13)?;
        let b = a.pow(3usize);
        let c = a.pow(-3isize);

        assert_eq!(b, FieldElement::new(5, 13)?);
        assert_eq!(c, FieldElement::new(8, 13)?);
        Ok(())
    }

    #[test]
    fn divition() -> Result<()> {
        let a = FieldElement::new(2, 19)?;
        let b = FieldElement::new(7, 19)?;
        let c = a.may_div(b)?;

        assert_eq!(c, FieldElement::new(3, 19)?);
        Ok(())
    }
}
