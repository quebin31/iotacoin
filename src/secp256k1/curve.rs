use std::ops::{Add, Mul};

use lazy_static::lazy_static;
use num_bigint::BigUint;
use num_traits::{One, Pow, Zero};

use crate::Error;

use super::field::FieldElement;
use super::field::PRIME;

lazy_static! {
    pub(crate) static ref ECURVE: EllipticCurve =
        EllipticCurve::new(FieldElement::new(0usize), FieldElement::new(7usize));
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct EllipticCurve {
    a: FieldElement,
    b: FieldElement,
}

impl EllipticCurve {
    pub fn new(a: FieldElement, b: FieldElement) -> Self {
        Self { a, b }
    }

    pub fn contains(&self, x: &FieldElement, y: &FieldElement) -> bool {
        y.pow(2usize) == x.pow(3usize) + &self.a * x + &self.b
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Point {
    AtInfinity,
    Normal(FieldElement, FieldElement),
}

impl Point {
    pub fn new(x: FieldElement, y: FieldElement) -> Result<Self, Error> {
        if ECURVE.contains(&x, &y) {
            Ok(Self::Normal(x, y))
        } else {
            Err(Error::PointNotOnTheCurve)
        }
    }

    pub fn at_infinity() -> Self {
        Self::AtInfinity
    }

    pub fn x(&self) -> Option<&FieldElement> {
        match self {
            Point::AtInfinity { .. } => None,
            Point::Normal(x, _) => Some(x),
        }
    }

    pub fn y(&self) -> Option<&FieldElement> {
        match self {
            Point::AtInfinity { .. } => None,
            Point::Normal(_, y) => Some(y),
        }
    }

    pub fn is_point_at_inf(&self) -> bool {
        matches!(self, Self::AtInfinity)
    }
}

impl Zero for Point {
    fn zero() -> Self {
        Point::AtInfinity
    }

    fn is_zero(&self) -> bool {
        matches!(self, Point::AtInfinity)
    }
}

impl<'a, 'b> Add<&'a Point> for &'b Point {
    type Output = Point;

    fn add(self, rhs: &'a Point) -> Self::Output {
        match (self, rhs) {
            // Additive identity
            (Point::AtInfinity, p) | (p, Point::AtInfinity) => p.to_owned(),

            // Normal addition between points
            (Point::Normal(x1, y1), Point::Normal(x2, y2)) => match (x1 == x2, y1 == y2) {
                // Same x axis, rhs is additive inverse of self and viceversa
                (true, false) => Point::at_infinity(),

                // Same x and y axis, self is equal to rhs
                (true, true) => {
                    if y1.is_zero() {
                        return Point::at_infinity();
                    }

                    let slope = (x1.pow(2usize) * 3usize + &ECURVE.a) / (y1 * 2usize);
                    let x3 = slope.pow(2usize) - (x1 * 2);
                    let y3 = slope * (x1 - &x3) - y1;

                    Point::Normal(x3, y3)
                }

                // Different x axis, y axis doesn't matter in this case
                _ => {
                    let slope = (y2 - y1) / (x2 - x1);
                    let x3 = slope.pow(2usize) - x1 - x2;
                    let y3 = slope * (x1 - &x3) - y1;

                    Point::Normal(x3, y3)
                }
            },
        }
    }
}

impl<'a, U> Mul<U> for &'a Point
where
    U: Into<BigUint>,
{
    type Output = Point;

    fn mul(self, coef: U) -> Self::Output {
        let mut coef = coef.into() % &*PRIME;

        let one = BigUint::one();
        let mut result = Point::zero();
        let mut current = self.clone();

        while !coef.is_zero() {
            if &coef & &one == one {
                result = &result + &current;
            }

            coef >>= 1;
            current = &current + &current;
        }

        result
    }
}

impl<U> Mul<U> for Point
where
    U: Into<BigUint>,
{
    type Output = Point;

    fn mul(self, coef: U) -> Self::Output {
        Mul::mul(&self, coef)
    }
}

forward_binop_impl!(for non-copyable Point where Add does add);
