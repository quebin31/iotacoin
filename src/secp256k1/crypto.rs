use hmac::{Hmac, Mac, NewMac};
use num_bigint::BigUint;
use num_traits::One;
use sha2::Sha256;

use crate::utils::{prepend_padding, ChainedMac};
use crate::Error;

use super::curve::Point;
use super::field::FieldElement;
use super::signature::Signature;
use super::{G, N};

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct PublicKey {
    pub(crate) ec_point: Point,
}

impl From<Point> for PublicKey {
    fn from(ec_point: Point) -> Self {
        Self { ec_point }
    }
}

impl PublicKey {
    pub fn new<U>(x: U, y: U) -> Result<Self, Error>
    where
        U: Into<BigUint>,
    {
        let x = FieldElement::new(x);
        let y = FieldElement::new(y);
        let ec_point = Point::new(x, y)?;

        Ok(Self { ec_point })
    }

    pub fn from_bytes_be<B>(x: B, y: B) -> Result<Self, Error>
    where
        B: AsRef<[u8]>,
    {
        let x = BigUint::from_bytes_be(x.as_ref());
        let y = BigUint::from_bytes_be(y.as_ref());
        Self::new(x, y)
    }

    pub fn from_bytes_le<B>(x: B, y: B) -> Result<Self, Error>
    where
        B: AsRef<[u8]>,
    {
        let x = BigUint::from_bytes_le(x.as_ref());
        let y = BigUint::from_bytes_le(y.as_ref());
        Self::new(x, y)
    }

    pub fn valid_signature<B>(&self, digest: B, signature: &Signature) -> Result<bool, Error>
    where
        B: AsRef<[u8]>,
    {
        signature.is_valid(digest, &self)
    }

    /// Serialize this public key using the SEC format
    pub fn serialize(&self, compressed: bool) -> Result<Vec<u8>, Error> {
        self.ec_point.serialize(compressed)
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct PrivateKey {
    pub(crate) secret: BigUint,
    pub(crate) pub_key: PublicKey,
}

impl PrivateKey {
    pub fn new<U>(secret: U) -> Self
    where
        U: Into<BigUint>,
    {
        let secret = secret.into();
        let ec_point = &*G * secret.clone();
        let pub_key = PublicKey { ec_point };

        Self { secret, pub_key }
    }

    pub fn from_bytes_be<B>(secret: B) -> Self
    where
        B: AsRef<[u8]>,
    {
        let secret = BigUint::from_bytes_be(secret.as_ref());
        Self::new(secret)
    }

    pub fn from_bytes_le<B>(secret: B) -> Self
    where
        B: AsRef<[u8]>,
    {
        let secret = BigUint::from_bytes_le(secret.as_ref());
        Self::new(secret)
    }

    pub fn public_key(&self) -> &PublicKey {
        &self.pub_key
    }

    pub fn create_signature<B>(&self, digest: B) -> Result<Signature, Error>
    where
        B: AsRef<[u8]>,
    {
        let digest = digest.as_ref();
        if digest.len() != 32 {
            return Err(Error::InvalidDigestLength(digest.len()));
        }

        let k = self.deterministic_k(digest)?;
        let r = (&*G * k.clone()).x().unwrap().0.clone();

        let k_inv = k.modpow(&(&*N - 2usize), &*N);
        let z = BigUint::from_bytes_be(digest);
        let mut s = (z + &r * &self.secret) * k_inv % &*N;
        if s > &*N / 2usize {
            s = &*N - s;
        }

        Ok(Signature::new(r, s))
    }

    fn deterministic_k<B>(&self, digest: B) -> Result<BigUint, Error>
    where
        B: AsRef<[u8]>,
    {
        type HmacSha256 = Hmac<Sha256>;

        let digest = digest.as_ref();
        debug_assert!(digest.len() == 32);

        let mut z = BigUint::from_bytes_be(digest);
        let k = [0x00u8; 32];
        let v = [0x01u8; 32];

        if z > *N {
            z -= &*N;
        }

        let secret_bytes = prepend_padding(self.secret.to_bytes_be(), 32, 0)?;

        let hmac = HmacSha256::new_varkey(&k).unwrap();
        let k = hmac
            .chain(&v)
            .chain(&[0x00])
            .chain(&secret_bytes)
            .chain(digest)
            .finalize()
            .into_bytes();

        let hmac = HmacSha256::new_varkey(&k).unwrap();
        let v = hmac.chain(&v).finalize().into_bytes();

        let hmac = HmacSha256::new_varkey(&k).unwrap();
        let mut k = hmac
            .chain(&v)
            .chain(&[0x01])
            .chain(&secret_bytes)
            .chain(digest)
            .finalize()
            .into_bytes();

        let hmac = HmacSha256::new_varkey(&k).unwrap();
        let mut v = hmac.chain(&v).finalize().into_bytes();

        let one = BigUint::one();
        loop {
            let hmac = HmacSha256::new_varkey(&k).unwrap();
            v = hmac.chain(&v).finalize().into_bytes();

            let candidate = BigUint::from_bytes_be(&v);
            if candidate >= one && candidate < *N {
                return Ok(candidate);
            }

            let hmac = HmacSha256::new_varkey(&k).unwrap();
            k = hmac.chain(&v).chain(&[0x00]).finalize().into_bytes();
            let hmac = HmacSha256::new_varkey(&k).unwrap();
            v = hmac.chain(&v).finalize().into_bytes();
        }
    }
}
