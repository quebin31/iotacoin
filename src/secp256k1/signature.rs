use std::io::Read;

use bytes::Buf;
use num_bigint::BigUint;

use crate::utils::strip_start;
use crate::Error;

use super::crypto::PublicKey;
use super::{G, N};

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Signature {
    pub(crate) r: BigUint,
    pub(crate) s: BigUint,
}

impl Signature {
    pub fn new<U>(r: U, s: U) -> Self
    where
        U: Into<BigUint>,
    {
        let r = r.into();
        let s = s.into();
        Self { r, s }
    }

    pub fn is_valid<B>(&self, digest: B, pub_key: &PublicKey) -> Result<bool, Error>
    where
        B: AsRef<[u8]>,
    {
        let digest = digest.as_ref();
        if digest.len() != 32 {
            return Err(Error::InvalidDigestLength(digest.len()));
        }

        let z = BigUint::from_bytes_be(digest);
        let s_inv = self.s.modpow(&(&*N - 2usize), &*N);

        let u = (&z * &s_inv) % &*N;
        let v = (&self.r * &s_inv) % &*N;

        let total = &*G * u + &pub_key.ec_point * v;
        Ok(total.x().unwrap().0 == self.r)
    }

    /// Serialize signature with DER format
    pub fn serialize(&self) -> Result<Vec<u8>, Error> {
        let r_bigendian = self.r.to_bytes_be();
        let r_bigendian = strip_start(&r_bigendian, 0x00);
        let r_bigendian = if r_bigendian[0] & 0x80 == 0x80 {
            std::iter::once(0x00u8)
                .chain(r_bigendian.iter().copied())
                .collect::<Vec<_>>()
        } else {
            r_bigendian.to_vec()
        };

        let prefix = [0x02u8, r_bigendian.len() as u8];
        let result = prefix.iter().copied().chain(r_bigendian);

        let s_bigendian = self.s.to_bytes_be();
        let s_bigendian = strip_start(&s_bigendian, 0x00);
        let s_bigendian = if s_bigendian[0] & 0x80 == 0x80 {
            std::iter::once(0x00u8)
                .chain(s_bigendian.iter().copied())
                .collect::<Vec<_>>()
        } else {
            s_bigendian.to_vec()
        };

        let result = result
            .chain([0x02u8, s_bigendian.len() as u8].iter().copied())
            .chain(s_bigendian)
            .collect::<Vec<_>>();

        let serialized = [0x30u8, result.len() as u8]
            .iter()
            .copied()
            .chain(result)
            .collect();

        Ok(serialized)
    }

    pub fn deserialize<B: Buf>(bytes: B) -> Result<Self, Error> {
        let size = bytes.remaining();
        let mut reader = bytes.reader();

        let mut buf = [0u8; 4];
        reader.read_exact(&mut buf)?;

        if buf[0] != 0x30 {
            return Err(Error::InvalidSignature("bad compound"));
        }

        let claimed_size = (buf[1] + 2) as usize;
        if claimed_size != size {
            return Err(Error::InvalidSignature("bad signature size"));
        }

        if buf[2] != 0x02 {
            return Err(Error::InvalidSignature("bad marker"));
        }

        let r_size = buf[3] as usize;
        let mut r_bytes = vec![0u8; r_size];
        reader.read_exact(&mut r_bytes)?;
        let r = BigUint::from_bytes_be(&r_bytes);

        let mut buf = [0u8; 2];
        reader.read_exact(&mut buf)?;

        if buf[0] != 0x02 {
            return Err(Error::InvalidSignature("bad marker"));
        }

        let s_size = buf[0] as usize;
        let mut s_bytes = vec![0u8; s_size];
        reader.read_exact(&mut s_bytes)?;
        let s = BigUint::from_bytes_be(&s_bytes);

        if size != 6 + r_size + s_size {
            return Err(Error::InvalidSignature("signature too long"));
        }

        Ok(Self { r, s })
    }
}

#[cfg(test)]
mod tests {
    use hex_literal::hex;

    use super::Signature;

    #[test]
    fn der_format() {
        let r = biguint!("37206a0610995c58074999cb9767b87af4c4978db68c06e8e6e81d282047a7c6");
        let s = biguint!("8ca63759c1157ebeaec0d03cecca119fc9a75bf8e6d0fa65c841c8e2738cdaec");
        let signature = Signature::new(r, s);

        let serialized = signature.serialize().unwrap();
        let expected = hex!(
            "3045022037206a0610995c58074999cb9767b87af4c4978db68c06e8e6e81d282047a7c6022100
            8ca63759c1157ebeaec0d03cecca119fc9a75bf8e6d0fa65c841c8e2738cdaec"
        );

        assert_eq!(serialized, expected);
    }
}
