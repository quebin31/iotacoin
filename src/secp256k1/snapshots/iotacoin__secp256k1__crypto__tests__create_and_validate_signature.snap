---
source: src/secp256k1/crypto.rs
expression: signature

---
Signature {
    r: BigUint {
        data: [
            11874953469058003620,
            16704089943962225621,
            5509877025105711965,
            15817134460319245706,
        ],
    },
    s: BigUint {
        data: [
            11592329467545065003,
            12950490178890555505,
            9395128666611346773,
            5984018298195652891,
        ],
    },
}
